src/plugins/dashboard/public/dashboard_actions/add_to_library_action.tsx

```diff
@@ -6,110 +6,70 @@
+    if (!isApiCompatible(embeddable)) throw new IncompatibleActionError();
     return 'folderCheck';
   }
 
-  public async isCompatible({ embeddable }: AddToLibraryActionContext) {
-    // TODO: Fix this, potentially by adding a 'canSave' function to embeddable interface
-    const { maps, visualize } = this.applicationCapabilities;
-    const canSave = embeddable.type === 'map' ? maps.save : visualize.save;
-    const { isOfAggregateQueryType } = await import('@kbn/es-query');
-    const query = isFilterableEmbeddable(embeddable) && (await embeddable.getQuery());
-    // Textbased panels (i.e. ES|QL, SQL) should not save to library
-    const isTextBasedEmbeddable = isOfAggregateQueryType(query as AggregateQuery);
-    return Boolean(
-      canSave &&
-        !isErrorEmbeddable(embeddable) &&
-        embeddable.getInput()?.viewMode !== ViewMode.VIEW &&
-        embeddable.getRoot() &&
-        embeddable.getRoot().isContainer &&
-        embeddable.getRoot().type === DASHBOARD_CONTAINER_TYPE &&
-        isReferenceOrValueEmbeddable(embeddable) &&
-        !embeddable.inputIsRefType(embeddable.getInput()) &&
-        !isTextBasedEmbeddable
-    );
+  public async isCompatible({ embeddable }: EmbeddableApiContext) {
+    if (!isApiCompatible(embeddable)) return false;
+    return embeddable.viewMode.value === 'edit' && (await embeddable.canLinkToLibrary());
   }
 
-  public async execute({ embeddable }: AddToLibraryActionContext) {
-    if (!isReferenceOrValueEmbeddable(embeddable)) {
-      throw new IncompatibleActionError();
-    }
-    const newInput = await embeddable.getInputAsRefType();
-
-    embeddable.updateInput(newInput);
-
-    const dashboard = embeddable.getRoot() as DashboardContainer;
-    const panelToReplace = dashboard.getInput().panels[embeddable.id] as DashboardPanelState;
-    if (!panelToReplace) {
-      throw new PanelNotFoundError();
-    }
-
-    const newPanel: PanelState<EmbeddableInput> = {
-      type: embeddable.type,
-      explicitInput: { ...newInput },
-    };
-    const replacedPanelId = await dashboard.replacePanel(panelToReplace, newPanel, true);
-
-    const title = dashboardAddToLibraryActionStrings.getSuccessMessage(
-      embeddable.getTitle() ? `'${embeddable.getTitle()}'` : ''
-    );
-
-    if (dashboard.getExpandedPanelId() !== undefined) {
-      dashboard.setExpandedPanelId(replacedPanelId);
+  public async execute({ embeddable }: EmbeddableApiContext) {
+    if (!isApiCompatible(embeddable)) throw new IncompatibleActionError();
+    try {
+      await embeddable.linkToLibrary();
+      const panelTitle = embeddable.panelTitle?.value ?? embeddable.defaultPanelTitle?.value;
+      this.toastsService.addSuccess({
+        title: dashboardAddToLibraryActionStrings.getSuccessMessage(
+          panelTitle ? `'${panelTitle}'` : ''
+        ),
+        'data-test-subj': 'addPanelToLibrarySuccess',
+      });
+    } catch (e) {
+      this.toastsService.addDanger({
```

Comment:
Should error message also include panel title?

Reply:
Good call, done in [b9f6ca7](https://github.com/elastic/kibana/pull/172017/commits/b9f6ca76d7c77d708667a7b2ec643d1ddbbb07db)